pkg/controllers/workgenerator/envelope.go

```diff
@@ -0,0 +1,359 @@
+			wrappedErr := fmt.Errorf("a namespaced object %s (%v) in has been wrapped in a resource envelope from another namespace (%v)", k, klog.KObj(uObj), envelopeReader.GetEnvelopeObjRef())
+			klog.ErrorS(wrappedErr, "Found an invalid manifest", "manifestKey", k, "envelope", envelopeReader.GetEnvelopeObjRef())
+			return nil, controller.NewUserError(wrappedErr)
+		}
+
+		manifests = append(manifests, fleetv1beta1.Manifest{
+			RawExtension: v.Data,
+		})
+	}
+
+	// Do a stable sort of the extracted manifests to ensure consistent, deterministic ordering.
+	sort.Slice(manifests, func(i, j int) bool {
+		obj1 := manifests[i].Raw
+		obj2 := manifests[j].Raw
+		// order by its json formatted string
+		return strings.Compare(string(obj1), string(obj2)) > 0
+	})
+	return manifests, nil
+}
+
+func refreshWorkForEnvelopeCR(
+	work *fleetv1beta1.Work,
+	resourceBinding *fleetv1beta1.ClusterResourceBinding,
+	resourceSnapshot *fleetv1beta1.ClusterResourceSnapshot,
+	manifests []fleetv1beta1.Manifest,
+	resourceOverrideSnapshotHash, clusterResourceOverrideSnapshotHash string,
+) {
+	// Update the parent resource snapshot index label.
+	work.Labels[fleetv1beta1.ParentResourceSnapshotIndexLabel] = resourceSnapshot.Labels[fleetv1beta1.ResourceIndexLabel]
+
+	// Update the annotations.
+	if work.Annotations == nil {
+		work.Annotations = make(map[string]string)
+	}
+	work.Annotations[fleetv1beta1.ParentResourceSnapshotNameAnnotation] = resourceBinding.Spec.ResourceSnapshotName
+	work.Annotations[fleetv1beta1.ParentResourceOverrideSnapshotHashAnnotation] = resourceOverrideSnapshotHash
+	work.Annotations[fleetv1beta1.ParentClusterResourceOverrideSnapshotHashAnnotation] = clusterResourceOverrideSnapshotHash
+
+	// Update the work spec (the manifests and the apply strategy).
+	work.Spec.Workload.Manifests = manifests
+	work.Spec.ApplyStrategy = resourceBinding.Spec.ApplyStrategy
+}
+
+func buildNewWorkForEnvelopeCR(
+	workNamePrefix string,
+	resourceBinding *fleetv1beta1.ClusterResourceBinding,
+	resourceSnapshot *fleetv1beta1.ClusterResourceSnapshot,
+	envelopeReader fleetv1alpha1.EnvelopeReader,
+	manifests []fleetv1beta1.Manifest,
+	resourceOverrideSnapshotHash, clusterResourceOverrideSnapshotHash string,
+) *fleetv1beta1.Work {
+	workName := fmt.Sprintf(fleetv1beta1.WorkNameWithEnvelopeCRFmt, workNamePrefix, uuid.NewUUID())
+	workNamespace := fmt.Sprintf(utils.NamespaceNameFormat, resourceBinding.Spec.TargetCluster)
+
+	return &fleetv1beta1.Work{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      workName,
+			Namespace: workNamespace,
+			Labels: map[string]string{
+				fleetv1beta1.ParentBindingLabel:               resourceBinding.Name,
+				fleetv1beta1.CRPTrackingLabel:                 resourceBinding.Labels[fleetv1beta1.CRPTrackingLabel],
+				fleetv1beta1.ParentResourceSnapshotIndexLabel: resourceSnapshot.Labels[fleetv1beta1.ResourceIndexLabel],
+				fleetv1beta1.EnvelopeTypeLabel:                string(fleetv1beta1.ConfigMapEnvelopeType),
```

Comment:
the matcher is using  envelopeReader.GetEnvelopeType() to find the work so this can't be configMap. However, does this type matter? CRD change will take a long time.